### DISCUSSION TITLE
Feasibility of Using Cake's Parser/AST Standalone

### DISCUSSION THREAD

**Discussion by scienceplease** [thradams/cake#188] [General]
@thradams, I'm curious what your opinion is w.r.t Cake's parser and AST representation being used in a standalone manner. I don't have enough familiarly with the structure of the source to make a good determination myself, but have been looking at other parsers/ASTs like the one in Clang or pycparser.

The end goal wouldn't be to do transpilation like Cake, but some some simple code analysis and transformations akin to what you would expect an LSP server to do.

Thanks!

**Comment by thradams** [comment]
Transformations can be done directly in cake creating a "visit" 
For instance, format_visit.c was created to format code.

It also possible to create a visit that generates the AST in some format, like json, or xml then consume 
the generated file in another project even in another language.

Cake needs a language server, is that you mean by LSP?

**Comment by scienceplease** [comment]
Thanks for the pointers! I'll focus on "format_visit.c" to start.

> It also possible to create a visit that generates the AST in some format, like json, or xml then consume
the generated file in another project even in another language.

Are you referring to the `-sarif` option? Can Cake read in these AST files back into it's internal representation?

> Cake needs a language server, is that you mean by LSP?

I did mean language server, but I should have clarified - the intention is to try and use Cake's C23 parser/AST to do simple code transformations on a codebase like refactor every usage of a function in a project.

**Reply by thradams** [reply]
>Are you referring to the -sarif option? Can Cake read in these AST files back into it's internal representation?
The option to generate json/xml does not exist. But the create one, it would be a "visit" on the AST.

>I did mean language server, but I should have clarified - the intention is to try and use Cake's C23 parser/AST to do simple code >transformations on a codebase like refactor every usage of a function in a project.

I think  Cake is a excellent choice for that, because it also does something similar.
In this case you also can see visit.c (this is the code that does the transformation C23->c99 for instance).
The idea of Cake transformations is we have the AST and we also have the list of tokens. 
Transformation are token insertion/edit and never token delete, instead cake hides tokens that should not appear. 
It can be used to create refactoring tools like rename.
